feat: add sprinkler support

FeatherKing opened the pull request:
This PR adds support for my irrigation api work in this other PR https://github.com/SecKatie/wyzeapy/pull/127 and related issue https://github.com/SecKatie/ha-wyzeapi/issues/230

Again, my first time adding to a home assistant component, but ive added my code to the sensor.py, switch.py and a new number.py.

Each irrigation device has multiple zones. The parent device has several sensors (RSSI, wifimac, ip address, etc), and then the zones are represented as switches. they have attributes as well (name, id, etc), but my thought in home assistant is to be able to turn them on and off like a switch. The number entity is how long the zone should run for (1-3600 seconds).

Happy to make corrections, but i need the api work merged before i can test much more here

FeatherKing commented on custom_components/wyzeapi/switch.py:
```diff
@@ -578,3 +590,86 @@ async def async_added_to_hass(self) -> None:
+        """Return the name of the zone."""
+        return f"{self._device.nickname} - {self._zone.name}"
+
+    @property
+    def is_on(self) -> bool:
+        """Return true if the zone is running."""
+        return self._is_running
```
i was thinking of removing this actually. There is no property from the Wyze api that shows if a zone is running. In fact, zones dont run, schedules run. The implementation is goofy on their side. a single "zone run" is just an adhoc schedule that runs on demand for a set duration. ill remove this here

brg468 replied:
So there's no way to see if a zone is currently running via the API?

FeatherKing replied:
nope. see the class attributes [here](https://github.com/SecKatie/wyzeapy/blob/6d373135e995aba902052186881fc902aff3e730/src/wyzeapy/services/irrigation_service.py#L56:L91)

brg468 replied:
Well if that's the case then I'm not sure you can use a switch. I think a switch has to have a state. Maybe a button platform would make more sense then?

FeatherKing replied:
oh ok. i can look into using a button for on (using the number payload), and a button for off

brg468 replied:
I guess the other option would be to create an assumed state using a switch and incorporating a timer. Aside from the complexity, another down side would be during a restart or when a zone is started from the app the switch state would always show off.

FeatherKing replied:
yeah that felt brittle to me so i was thinking to just remove. ill update this to a button and once everything is merged, do some real world testing with it

brg468 replied:
If you’re not aware already, you can copy your updated version of the wyzeapy folder into the config directory of HA (same location as your configuration.yaml file) and have this updated version in custom_components and that allows it to run the updated dependency and you should be able to test everything at that point. There’s some other ways to use custom dependencies but this is the easiest.

FeatherKing replied:
oh awesome. ill try that.

FeatherKing commented on custom_components/wyzeapi/switch.py:
```diff
@@ -578,3 +590,86 @@ async def async_added_to_hass(self) -> None:
+
+    async def async_turn_on(self, **kwargs: Any) -> None:
+        """Turn the zone on."""
+        await self._service._irrigation_start_zone(
```
makes sense, i was combining a lot of snippets from different places to see how things were implemented. ill clean up the other library as well

FeatherKing commented:
thanks for the review. i agree on reducing the sensors. In fact, i commented out a ton of "zone" properties in the other library for this same reason. Ill address all this feedback tonight.

FeatherKing commented on custom_components/wyzeapi/switch.py:
```diff
@@ -578,3 +590,86 @@ async def async_added_to_hass(self) -> None:
+        return f"{self._device.mac}-zone-{self._zone.zone_number}"
+
+    @property
+    def device_info(self) -> DeviceInfo:
```
im no longer modifying the switch.py file. do you want me to update this for the other functions in here? or i can make a separate PR for that

brg468 replied:
You can leave it as is to keep this PR on track. We’ll get to it eventually 😛

FeatherKing commented on custom_components/wyzeapi/switch.py:
```diff
@@ -578,3 +590,86 @@ async def async_added_to_hass(self) -> None:
+        self._is_running = False
+        self.async_write_ha_state()
+
+    async def async_added_to_hass(self) -> None:
+        """Subscribe to updates."""
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass,
+                f"{DOMAIN}-irrigation-{self._device.mac}",
+                self._handle_update,
+            )
+        )
```
does every single class need this callback? in the case of sensors, every single sensor needs this to update itself?

brg468 replied:
Short answer no, not every entity needs this. Basically the way its been done in this lib is to have the main device register the updater, which is then added to the custom update manager that the wyzeapy lib uses to control polling. To avoid having a ton of things in the update que we just add one updated that should be able to update all the related entities. So in this case, it looks like all the sensors can be updated by adding that callback to the irrigation sensor base class like you did. 

I'm not 100%, but quick glance through the other lib it looks like the quick run duration for the number entity and the sensor data are all updated at the same time, so having the sensor send out a dispatcher signal to the number entity that it's been updated and then have a listener in number to run its update method then. A good example of this is look at the WyzeSwitch class which registers an updater and sends a dispatcher signal, and then the WyzeCameraNotificationSwitch right below it that receives the dispatcher signals and updates from there. By default most of our updates are set at 30 second intervals, but since there's very little that would likely change often a higher update interval could probably be used.

As far as the button, buttons are stateless so they don't have anything to be updated, so there doesn't need to be an update method there.

Hope that makes some sense.

FeatherKing replied:
i actually realized i dont need state for the number entity either. I am using the number field in HA simply to give a way to POST that value to the quickrun api from Wyze. There is no quickrun value from the wyze api.

FeatherKing commented:
i believe i addressed all the feedback so far. i still have a question on some of the callbacks, specifically in the sensors.py. I think the rest of the callbacks looked ok. I needed to add a couple functions to the api to support what i was doing here. Sorry for the refactor of device and service, that made a bunch of line changes. I cant tell whats HA convention so irrigation instead of device seemed a little more clear.

brg468 commented on custom_components/wyzeapi/sensor.py:
```diff
@@ -499,152 +498,130 @@ async def async_added_to_hass(self) -> None:
-        """Return the SSID."""
-        return self._device.ssid
+        """Return the serial number."""
+        return self._irrigation.sn
```
This can be eliminated by just adding `serial_number=self._irrigation.serial` to DeviceInfo in the base class

brg468 commented on custom_components/wyzeapi/button.py:
```diff
@@ -0,0 +1,145 @@
+            _LOGGER.error("Error starting zone: %s", err)
+            raise HomeAssistantError(f"Failed to start zone: {err}") from err
+
+    @callback
+    def async_update_callback(self, irrigation: IrrigationDevice) -> None:
+        """Update the button's state."""
+        self._irrigation = irrigation
+        # Find the updated zone
+        for zone in irrigation.zones:
+            if zone.zone_number == self._zone.zone_number:
+                self._zone = zone
+                break
+        self.schedule_update_ha_state()
+
+    async def async_added_to_hass(self) -> None:
+        """Subscribe to updates."""
+        self._irrigation.callback_function = self.async_update_callback
+        self._irrigation_service.register_updater(self._irrigation, 30)
+        await self._irrigation_service.start_update_manager()
+        return await super().async_added_to_hass()
+
+    async def async_will_remove_from_hass(self) -> None:
+        """Clean up when removed."""
+        self._irrigation_service.unregister_updater(self._irrigation) 
```
See my comment about updaters, but this is not needed since a button is stateless.

brg468 commented on custom_components/wyzeapi/button.py:
```diff
@@ -0,0 +1,145 @@
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID for the zone."""
+        return f"{self._irrigation.mac}-zone-{self._zone.zone_number}"
```
Maybe "Start" at the beginning of the name so its clear what the button does.

brg468 replied:
```suggestion
        return f"Start {self._irrigation.mac}-zone-{self._zone.zone_number}"
```

FeatherKing commented:
I addressed all your feedback, thanks! I also took your tip of manually loading it into Home Assistant. I cleaned up names and icons and it seems like everything is updating and persisting as i expect. Ive restarted HA and im not seeing any surprises.

The only problem left is that pushing the buttons (my POST payloads back to wyze) are failing with a 'Signature2 is not valid'. I need to investigate that tomorrow, but the payload from HA shouldnt change so this might be done.

<img width="676" alt="image" src="https://github.com/user-attachments/assets/ebd34996-19e7-4b71-946f-c1603980e7a6" />

brg468 commented on custom_components/wyzeapi/number.py:
```diff
@@ -0,0 +1,176 @@
+
+    async def async_added_to_hass(self) -> None:
+        """Subscribe to updates."""
+        await self.async_get_last_number_data()
```
Isn't this redundant with L127? Seems like you can just call `_async_load_value `and that will get the last number data.

FeatherKing replied:
ill check this and `async_will_remove_from_hass`. i thought in my testing they were required, but that might have been before i moved to using a `RestoreNumber`

FeatherKing replied:
yeah this was a dupe from above

brg468 reviewed (commented):
Just a couple more minor things then I think I'm done 😄. And can you remove the comment out code at the end of number.py.

As far as the signature error, I have no idea where we get the signature for or why that would be throwing an error. Hopefully you can figure that out or @SecKatie may have some better insight into that.

FeatherKing commented on custom_components/wyzeapi/number.py:
```diff
@@ -0,0 +1,176 @@
+    @property
+    def native_max_value(self) -> float:
+        """Return the maximum value."""
+        return 3600.0  # 1 hour
```
no, but i was thinking that i wouldnt want to accidently run my sprinklers forever. I was thinking about some sane upper limit. i dont think there is a limit on the Wyze side (havent tested)

brg468 replied:
Ok, I just know from working on Rachio that there were a surprising number of people that run zones for hours sometimes for drip and trees so maybe an hour would be a little short. That said I don't know what the perfect answer is, but for reference Rachio's limit is 3 hours.

FeatherKing commented on custom_components/wyzeapi/number.py:
```diff
@@ -0,0 +1,176 @@
+
+    async def async_will_remove_from_hass(self) -> None:
+        """Clean up when removed."""
+        pass
```
yep, works fine without it now

FeatherKing commented:
thanks again! i addressed all the feedback. I think we should not merge until i can figure out the Signature2 error. Ive seen how it makes the payload, i just dont understand why its failing yet. Lets leave this open in case i need to combe back and make changes here.

FeatherKing commented:
fixed the signature errors in the other lib, but found one bug with the quickrun duration. sprinklers are turning on though!!

FeatherKing commented:
ok, this is working for me. Final bug was to make the button press use the home assistant number entity value instead of trying to use the Zone class attribute. @brg468 please review that logic because im not sure if there is a better way to get the number value from the button platform.

What i have implemented here is working, i just tested several zones and i can watch the changes appear in my phone app. so its "working" just want another pair of eyes on calling between platforms.

FeatherKing commented:
also, i wrote a home assistant automation now to run through all the zones once and then again (so the water can soak into the ground) and the automation will finish everything by sunrise. I could add that here as a reference file or maybe in the wiki.

i do miss my Rachio, but this is what we had at the new house. Its working for me now though, ran through a full cycle.

brg468 commented:
> ok, this is working for me. Final bug was to make the button press use the home assistant number entity value instead of trying to use the Zone class attribute. @brg468 please review that logic because im not sure if there is a better way to get the number value from the button platform.
> 
> What i have implemented here is working, i just tested several zones and i can watch the changes appear in my phone app. so its "working" just want another pair of eyes on calling between platforms.

Good to hear its working! Let me think about an easier way we could pass the number to the button, what you have works but it seems overly complicated.

brg468 commented:
So why can't you just pass the value that is set in Zone `quickrun_duration` during the button press? That should be set by the number entity right?

FeatherKing commented:
to be clear, the `quickrun_duration` is not a wyzeapi stored parameter at all. It takes `duration` as a payload when you call `quickrun`. In this case, i was using home assistant as the sole storage for `quickrun_duration`. It gets assigned into the Button as an extended attribute because I originally figured it could just be a property of the button. From what i read, the extended attributes are meant to be read-only and there isnt a convenient way to have a user update them.

So then i created a number entity, and when you pushed the button, it pulls the value from the number entity and sends that.

maybe ideally, there is no number entity, and somehow the user could update that attribute.

brg468 commented:
But isn't this assigning a value to the `quickrun_duration` variable of the Zone class? Which could then be accessed by the button platform? 

Sorry its tough to be exactly sure without having something to test with. Any chance you'd be up for sharing the device with my account so I could actually see it in action? I'd change the output so it wouldn't actually turn anything on..

```
async def async_set_native_value(self, value: float) -> None:
        """Set the value."""
        await self._irrigation_service.set_zone_quickrun_duration(
            self._device, self._zone.zone_number, int(value)
        )
        self._zone.quickrun_duration = int(value)
        self.async_write_ha_state()
```

FeatherKing commented:
i think when i tried that it didnt work (back when the button had state. but i may have misunderstood state versus the extended attributes at that time. Let me try again now that ive spent more time in the code. if i cant make it work ill share with you. i have a disabled zone that you could test starting and it wont flood my yard 😂

FeatherKing commented:
looking closer, i think the issue is that the zone class is a different instance. the button platform gets its own irrigation device

`button.py`
>     irrigation_service = await client.irrigation_service
>
>     # Get all irrigation devices
>     irrigation_devices = await irrigation_service.get_irrigations()
>     # Create a button entity for each zone in each irrigation device
>     buttons = []
>     for device in irrigation_devices:
>         # Update the device to get its zones
>         device = await irrigation_service.update(device)
>         for zone in device.zones:
>             if zone.enabled:
>                 buttons.append(WyzeIrrigationZoneButton(irrigation_service, device, zone))
>         # Add a stop all schedules button for each irrigation device, not each zone
>         buttons.append(WyzeIrrigationStopAllButton(irrigation_service, device))

`number.py`

>     irrigation_service = await client.irrigation_service
> 
>     # Get all irrigation devices
>     irrigation_devices = await irrigation_service.get_irrigations()
> 
>     # Create a number entity for each zone in each irrigation device
>     entities = []
>     for device in irrigation_devices:
>         # Update the device to get its zones
>         device = await irrigation_service.update(device)
>         for zone in device.zones:
>             if zone.enabled:
>                 entities.append(WyzeIrrigationQuickrunDuration(irrigation_service, device, zone))

since the wyze api doesnt actually have the value, the calls to irrigation_service.update() never update the device. whatever they pass in gets passed back. The number entity does update self._zone.quickstart_duration, but its updating its own copy of it.

Is there another way to keep a global state within HA? It seems like a weird scenario with the wyze api. I could make a totally new parent device type in HA but that feels more complicated that just having the button look at the number entity value, or having the number entity call a button service to update its attribute.

FeatherKing commented:
maybe the right way is to create a top level device. This is unique because its a device with several different sensor types. currently i am building some buttons, numbers and sensors and relying on HA to link them together at the device level. It feels closer to a climate platform (set temp, sensor for temp, switches for actions) but climate has that built in.

Let me look around for some other HA examples of a device that might exposes multiple platform types

brg468 commented:
Did you make any progress on this?

FeatherKing commented:
I didnt find a great way to build an abstraction so ive just been using it as designed so far in this PR. There might be one or two small changes that I need to commit here, but its been running for a month and half with no issues. I still havent received any reviews on the api PR. Ill push the changes here later today to make it up to date with what ive been running

FeatherKing commented:
this PR now matches what ive been running at home for two months or so

FeatherKing commented:
@SecKatie let me know what to do (if anything in this PR). I also have some templates that have been helpful for running the sprinklers at my house. I dont see a place for documentation really. i could add the yaml as files in this repo, or not include them at all.

brg468 commented on custom_components/wyzeapi/button.py:
```diff
@@ -0,0 +1,275 @@
+
+            # Find number entities for this device
+            number_entities = [
+                entity_id
+                for entity_id, entity in entity_registry.entities.items()
+                if entity.device_id == device.id
+                and entity.platform == DOMAIN
+                and entity_id.startswith("number.")
+            ]
+            _LOGGER.debug(f"Found number entities for device {self._device.mac}: {number_entities}")
+
+            # Find the matching number entity
+            matching_entity = None
+            # Normalize zone name to be a valid Home Assistant entity name
+            zone_name_normalized = ''.join(c if c.isalnum() else '_' for c in self._zone.name.lower())
+            # Remove double underscores
+            while '__' in zone_name_normalized:
+                zone_name_normalized = zone_name_normalized.replace('__', '_')
+            # Remove leading/trailing underscores
+            zone_name_normalized = zone_name_normalized.strip('_')
+            
+            for entity_id in number_entities:
+                entity = entity_registry.entities[entity_id]
+                # Extract the zone part from entity_id (e.g., 'backyard_s' from 'number.hag_controller_backyard_s')
+                entity_id_suffix = entity_id.split(".")[-1].lower()  # e.g., 'hag_controller_backyard_s' -> 'backyard_s'
+                # Also check unique_id and original_name
+                unique_id = entity.unique_id.lower() if entity.unique_id else ""
+                original_name = entity.original_name.lower() if entity.original_name else ""
+                _LOGGER.debug(
+                    f"Checking entity {entity_id}: "
+                    f"suffix={entity_id_suffix}, unique_id={unique_id}, original_name={original_name}, "
+                    f"zone_normalized={zone_name_normalized}"
+                )
+                if (
+                    entity_id_suffix.endswith(zone_name_normalized)
+                    or unique_id.endswith(zone_name_normalized)
+                    or original_name == zone_name_normalized
+                ):
+                    matching_entity = entity_id
+                    break
```
Can you test and see if this or something similar works? I think the less matching we have to do the better we’ll be for oddly named zones.

SecKatie replied:
Just let me know what your testing shows. This seems like a good suggestion and is much cleaner.

FeatherKing replied:
sorry, july has been crazy at work. i used the code above as a start of something cleaner. testing looks okay. let me run it overnight and see if it goes through both of my controllers and all zones.

SecKatie commented:
> @SecKatie let me know what to do (if anything in this PR). I also have some templates that have been helpful for running the sprinklers at my house. I dont see a place for documentation really. i could add the yaml as files in this repo, or not include them at all.

@FeatherKing This PR is looking really good. I am thinking once we resolve that one copilot suggestion this will be good to go into the code base.

brg468 commented:
The new changes look good to me 👍

SecKatie reviewed (approved):
Looks good to me
